impl/asm/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java

```diff
@@ -250,9 +261,63 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
                         mv.visitMethodInsn(opcode, owner, name, desc, itf);
                     }
 
-                    private void generateReflectionHandlingModelCode(String owner) {
-                        System.out.println("TODO: check if we generated reflection handling code in package " + owner + " yet, if not, generate it." +
-                                "We can track the (TransformerImpl level) Set of packages that we have already generated the code for, so we know if we need to generate.");
+                    private void generateReflectionHandlingModelCode(String handlingClassName) {
+                        // check if we generated reflection handling code yet, if not, generate it
+                        if (!generatedReflectionModelHandlingCode.contains(handlingClassName)) {
```

Comment:
Because of concurrency contains() call should be add() method call
with boolean return value inspection if the add operation was successful or not.
(of course with precondition HashSet was replaced with CopyOnWriteArraySet as suggested before).

Reply:
fixed in updates just pushed that use CopyOnWriteArraySet, good idea!

Reply:
If two threads come in and call contains() at same time, they both might see false and both generate extra class.  Probably need to add synchronized lock for that.

Reply:
Or we could create the CopyOnWriteArraySet with a class Handler { void handle(); ... } that does the extra class bytecode generation.

Reply:
Not sure about Handler class shown in CopyOnWriteArraySet, will do the synchronized way.

Reply:
should be fine with synchronized block, pushing that now.

Reply:
Concurrency is now fixed.